### PR TITLE
Fix FEInterfaceValues::reinit when no dominating FE has been found.

### LINKED ISSUE #15021: FEInterfaceValues::reinit fails when neither element dominates
As the title says. The following is an example (adapted from existing tests of FEInterfaceValues).
```c++

// Set up a two cells with different elements, such that neither element
// dominates. Create an FEInterfaceValues object and call reinit on it.

#include <deal.II/base/quadrature_lib.h>

#include <deal.II/fe/fe_interface_values.h>
#include <deal.II/fe/fe_q.h>
#include <deal.II/fe/fe_system.h>

#include <deal.II/grid/grid_generator.h>
#include <deal.II/grid/tria.h>

#include "../tests.h"

template <int dim>
void
make_2_cells(Triangulation<dim> &tria);

template <>
void
make_2_cells<2>(Triangulation<2> &tria)
{
  const unsigned int        dim         = 2;
  std::vector<unsigned int> repetitions = {2, 1};
  Point<dim>                p1;
  Point<dim>                p2(2.0, 1.0);

  GridGenerator::subdivided_hyper_rectangle(tria, repetitions, p1, p2);
}


template <int dim>
void
test()
{
  Triangulation<dim> tria;
  make_2_cells(tria);

  const unsigned      max_fe_degree = 2;
  const FESystem<dim> fe_left(FE_Q<dim>(1), FE_Q<dim>(2));
  const FESystem<dim> fe_right(FE_Q<dim>(2), FE_Q<dim>(1));

  hp::FECollection<dim> fe_collection;
  fe_collection.push_back(fe_left);
  fe_collection.push_back(fe_right);

  DoFHandler<dim> dofh(tria);
  dofh.distribute_dofs(fe_collection);

  const QGauss<dim - 1>          quadrature(max_fe_degree + 1);
  const hp::QCollection<dim - 1> q_collection(quadrature);


  const unsigned int face_index = 1;

  auto cell     = dofh.begin();
  auto neighbor = cell->neighbor(face_index);

  cell->set_active_fe_index(0);
  neighbor->set_active_fe_index(1);

  const UpdateFlags update_flags = update_values | update_gradients |
                                   update_quadrature_points | update_JxW_values;
  FEInterfaceValues<dim> fiv(fe_collection, q_collection, update_flags);
  fiv.reinit(cell,
             face_index,
             numbers::invalid_unsigned_int,
             neighbor,
             cell->neighbor_of_neighbor(face_index),
             numbers::invalid_unsigned_int);
}



int
main()
{
  initlog();
  test<2>();
}
```

Gives the error:
```
--------------------------------------------------------
An error occurred in line <456> of file </home/simon/workspace/dealii3/source/hp/fe_values.cc> in function
    void dealii::hp::FEFaceValues<dim, spacedim>::reinit(const dealii::TriaIterator<dealii::DoFCellAccessor<dim, spacedim, lda> >&, unsigned int, unsigned int, unsigned int, unsigned int) [with bool lda = false; int dim = 2; int spacedim = 2]
The violated condition was: 
    ::dealii::deal_II_exceptions::internals::compare_less_than(real_q_index, this->q_collections.size())
Additional information: 
    Index 65535 is not in the half-open range [0,1).
--------------------------------------------------------
```

The problem is on line 2494 in `fe_interface_values.h`. `find_dominated_fe` returns `numbers::invalid_fe_index` when neither element dominates.

```c++
const unsigned int dominated_fe_index =
        internal_hp_fe_face_values->get_fe_collection().find_dominated_fe(
          {cell->active_fe_index(), cell_neighbor->active_fe_index()});

      const unsigned int used_q_index =
        (q_index == numbers::invalid_unsigned_int ? dominated_fe_index :
                                                    q_index);
      const unsigned int used_mapping_index =
        (mapping_index == numbers::invalid_unsigned_int ? dominated_fe_index :
                                                          mapping_index);
```

How should we actually choose `q_index` and `mapping_index` when no element dominates? I could not figure it out by myself.

### PR DESCRIPTION
**Pull request by marcfehling**
Fixes #15021. Related to https://github.com/dealii/dealii/pull/14547#discussion_r1053319521.

~I cleaned up the `make_2_cells` functions first as we have a function `hyper_line` that does exactly the same.~ EDIT: Now part of #15037.

Next comes the actual fix to the problem of neighboring FEs that do not dominate each other. I used the test that @simonsticko provided, in which I noticed that `distribute_dofs()` was called at the wrong place. Luckily this throws an error, so I improved the error message.

@simonsticko @jppelteret @fdrmrc -- FYI

### CODE REVIEW

**Comment by marcfehling** [comment]
@bangerth -- Officially, the [`find_dominated_fe`](https://www.dealii.org/developer/doxygen/deal.II/classhp_1_1FECollection.html#a719793132cab791f215ee96b4d5f0c75) function still returns an `unsigned int`. I simply didn't make it so far in #13595 yet. Let's leave it this way for now.

https://github.com/dealii/dealii/blob/0cf9555689bad78df6ec9086a7bba3d82413d182/include/deal.II/fe/fe_interface_values.h#L2494-L2496

**Review by bangerth — CHANGES_REQUESTED** [review]
I don't think this is right yet.

**Comment by bangerth** [comment]
The first commit is of course correct in itself. Feel free to submit it separately.

File: include/deal.II/fe/fe_interface_values.h [hunk 1]
```diff
@@ -2496,11 +2496,22 @@ FEInterfaceValues<dim, spacedim>::reinit(
...
       const unsigned int used_q_index =
-        (q_index == numbers::invalid_unsigned_int ? dominated_fe_index :
-                                                    q_index);
+        [q_index, dominated_fe_index, &cell]() -> unsigned int {
+        if (q_index != numbers::invalid_unsigned_int)
+          return q_index;
+        if (dominated_fe_index != numbers::invalid_fe_index)
+          return dominated_fe_index;
+        return cell->active_fe_index();
+      }();
```
Comment (bangerth) [inline include/deal.II/fe/fe_interface_values.h:2505]: As stated elsewhere, I don't think this is right. We should be choosing the same quadrature formula regardless of which side of a face you approach.

**Pull request closed without merging**